File: Web/packages/kernel/test/logs-conformance.test.ts
//
//  logs-conformance.test.ts - the SHARED log corpus (OpenSource/Conformance/logs/logs.json)
//  through the TS kernel. The Kotlin (LogsConformanceTest.kt) and Swift
//  (ConformanceHosts.LogsConformance) twins run the SAME file — fixtures-first for the
//  unified console primitive: `dsx.log` house formatting + redaction, source-scheme
//  attribution (markup / module handle / console.* / the kernel `dsx.log` bus verb),
//  and the log ring (cap 500).
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import { ReactiveStore } from "../src/store.ts";
import { ActionRunner, makeRunEnv } from "../src/runner.ts";
import { ModuleRegistry, ModuleCallError, makeDsx } from "../src/bus.ts";
import { DSXLogs } from "../src/logs.ts";
import { JSE } from "../src/jse/jse.ts";
import { NSNull, isDict, string, type Dict } from "../src/jse/values.ts";

function corpusFile(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/logs/logs.json");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("logs corpus not found");
    dir = parent;
  }
}

type Case = {
  name: string;
  steps: Array<Dict>;
  expect: {
    logs?: Dict[];
    logCount?: number;
    logsTail?: Dict[];
    callErrors?: string[];
    jseStore?: { [path: string]: unknown };
  };
};

/** SUBSET match — the errors-corpus discipline: listed keys deep-match, unlisted ignored. */
function subset(actual: unknown, expected: unknown): boolean {
  if (expected === null) return actual === null || actual === undefined || actual === NSNull;
  if (Array.isArray(expected)) {
    if (!Array.isArray(actual) || actual.length !== expected.length) return false;
    return expected.every((e, i) => subset(actual[i], e));
  }
  if (typeof expected === "object") {
    if (actual === null || typeof actual !== "object") return false;
    return Object.entries(expected as Dict).every(([k, v]) => subset((actual as Dict)[k], v));
  }
  return actual === expected;
}

function assertSubset(actual: unknown, expected: unknown, label: string): void {
  assert.ok(subset(actual, expected),
    `${label}: expected subset ${JSON.stringify(expected)} — got ${JSON.stringify(actual)}`);
}

const doc = JSON.parse(readFileSync(corpusFile(), "utf-8")) as { cases: Case[] };
assert.ok(doc.cases.length > 0, "logs corpus is empty");

for (const c of doc.cases) {
  test(`logs-corpus/${c.name}`, async () => {
    const countBefore = DSXLogs.count();
    const callErrors: string[] = [];
    let jseStore: ReactiveStore | null = null;

    for (const step of c.steps) {
      if (isDict(step["jse"])) {
        const j = step["jse"] as Dict;
        jseStore = new ReactiveStore();
        const env = makeRunEnv(jseStore, { ownerScheme: j["scheme"] === undefined ? undefined : string(j["scheme"]) });
        await new ActionRunner(env).run(string(j["body"]));
      } else if (isDict(step["log"])) {
        const l = step["log"] as Dict;
        const args = Array.isArray(l["args"]) ? (l["args"] as unknown[]) : [];
        makeDsx(string(l["scheme"])).log(...args);
      } else if (isDict(step["logRepeat"])) {
        const r = step["logRepeat"] as Dict;
        const dsx = makeDsx(string(r["scheme"]));
        const count = Number(r["count"]);
        for (let i = 0; i < count; i++) dsx.log(`${string(r["prefix"])}${i}`);
      } else if (isDict(step["call"])) {
        const k = step["call"] as Dict;
        const args = isDict(k["args"]) ? (k["args"] as Dict) : {};
        try {
          await ModuleRegistry.dispatch(string(k["scheme"]), string(k["action"]), args);
        } catch (e) {
          callErrors.push(e instanceof ModuleCallError ? e.code : "error");
        }
      }
    }

    if (c.expect.logs) {
      const appended = DSXLogs.count() - countBefore;
      assert.equal(appended, c.expect.logs.length, `${c.name}: appended log entries`);
      const tail = DSXLogs.recent().slice(-appended);
      c.expect.logs.forEach((exp, i) => assertSubset(tail[i], exp, `${c.name}: logs[${i}]`));
    }
    if (c.expect.logCount !== undefined) {
      assert.equal(DSXLogs.recent().length, c.expect.logCount, `${c.name}: retained log total`);
    }
    for (const [i, exp] of (c.expect.logsTail ?? []).entries()) {
      const tail = DSXLogs.recent().slice(-(c.expect.logsTail!.length));
      assertSubset(tail[i], exp, `${c.name}: logsTail[${i}]`);
    }
    if (c.expect.callErrors) assert.deepEqual(callErrors, c.expect.callErrors, `${c.name}: callErrors`);
    for (const [path, exp] of Object.entries(c.expect.jseStore ?? {})) {
      assert.ok(jseStore, `${c.name}: no jse step ran`);
      const actual = JSE.eval(path, jseStore!.jse, null);
      assertSubset(actual === NSNull ? null : actual, exp, `${c.name}: jseStore ${path}`);
    }
  });
}

File: Engine/Android/core/src/test/kotlin/despia/engine/LogsConformanceTest.kt
package despia.engine

import java.io.File
import java.util.concurrent.Executor
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The log-corpus conformance runner — executes OpenSource/Conformance/logs/logs.json
 * through THIS runtime's statement runner + log ring (the TS runner and the Swift
 * reference run the SAME file). The law under test (logs/README.md): `dsx.log` house
 * formatting + redaction, source-scheme attribution (markup / module handle / console.* /
 * the kernel `dsx.log` bus verb behind the page's `dsx.log`), and the log ring (cap 500).
 */
class LogsConformanceTest {

    private fun corpusFile(): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/logs/logs.json")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/logs/logs.json not found")
        }
    }

    private companion object {
        /** Constructor stash — the ErrorsConformanceTest discipline (Module.<init> resolves
         *  `scheme` before subclass properties assign; tests are single-threaded). */
        var pendingScheme = ""
        val emitters = HashMap<String, Module>()
    }

    private class LogModule : Module() {
        private var schemeName: String? = null
        override val scheme get() = schemeName ?: pendingScheme
        init { schemeName = pendingScheme }
        override fun setup() {}
    }

    /** A bare module whose dsx IS the module handle for `scheme` — the module-handle
     *  `dsx.log` form needs no registered actions, only a primary scheme. */
    private fun handleDsx(scheme: String): Context {
        emitters[scheme]?.let { return it.dsx }
        pendingScheme = scheme
        val mod = LogModule()
        ModuleRegistry.shared.register { mod }
        emitters[scheme] = mod
        return mod.dsx
    }

    private fun subset(actual: Any?, expected: Any?): Boolean = when (expected) {
        null, NSNull -> actual == null || actual == NSNull
        is Map<*, *> -> (actual as? Map<*, *>)?.let { a ->
            expected.entries.all { (k, v) -> subset(a[k], v) }
        } ?: false
        is List<*> -> (actual as? List<*>)?.takeIf { it.size == expected.size }
            ?.let { a -> expected.indices.all { subset(a[it], expected[it]) } } ?: false
        else -> JSE.equals(actual, expected)
    }

    private fun assertSubset(actual: Any?, expected: Any?, label: String) {
        assertTrue(subset(actual, expected), "$label: expected subset $expected — got $actual")
    }

    @TestFactory
    fun logsCorpus(): List<DynamicTest> {
        val doc = json(corpusFile().readText()).foundationValue as? Map<*, *>
            ?: error("logs.json: not a JSON object")
        @Suppress("UNCHECKED_CAST")
        val cases = doc["cases"] as? List<Map<String, Any?>> ?: error("logs.json: no cases[]")
        return cases.map { c ->
            val name = "logs-corpus/${c["name"]}"
            DynamicTest.dynamicTest(name) { runCase(name, c) }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun runCase(name: String, c: Map<String, Any?>) {
        ModuleRegistry.shared.mainExecutor = Executor { it.run() }
        JSERunner.mainExecutor = Executor { it.run() }
        JSERunner.backgroundExecutor = Executor { it.run() }

        val expect = (c["expect"] as? Map<String, Any?>) ?: emptyMap()
        val countBefore = DSXLogBuffer.shared.count()
        val callErrors = ArrayList<String>()
        var jseStore: StackStore? = null

        for (step in (c["steps"] as? List<Map<String, Any?>>) ?: emptyList()) {
            (step["jse"] as? Map<String, Any?>)?.let { j ->
                val store = StackStore()
                jseStore = store
                store.registerAction("corpuscase", emptyMap(), JSE.string(j["body"]))
                val runner = JSERunner(store, scope = j["scheme"] as? String)
                runner.run("corpuscase", null)
            }
            (step["log"] as? Map<String, Any?>)?.let { l ->
                val args = (l["args"] as? List<Any?>) ?: emptyList()
                handleDsx(JSE.string(l["scheme"])).log(*args.toTypedArray())
            }
            (step["logRepeat"] as? Map<String, Any?>)?.let { r ->
                val dsx = handleDsx(JSE.string(r["scheme"]))
                val count = (JSE.number(r["count"]) ?: 0.0).toInt()
                val prefix = JSE.string(r["prefix"])
                repeat(count) { i -> dsx.log("$prefix$i") }
            }
            (step["call"] as? Map<String, Any?>)?.let { k ->
                val callArgs = k["args"] as? Map<String, Any?>
                val caller = Module().dsx
                runBlocking {
                    try { caller.module[JSE.string(k["scheme"])][JSE.string(k["action"])](callArgs) }
                    catch (e: ModuleCallError) {
                        callErrors.add(when (e) {
                            is ModuleCallError.ActionFailed -> e.code
                            is ModuleCallError.NotLoaded -> "not_loaded"
                            is ModuleCallError.InvalidURI -> "invalid_uri"
                        })
                    }
                }
            }
        }

        (expect["logs"] as? List<Map<String, Any?>>)?.let { exp ->
            val appended = DSXLogBuffer.shared.count() - countBefore
            assertEquals(exp.size, appended, "$name: appended log entries")
            val tail = DSXLogBuffer.shared.recent().takeLast(appended)
            exp.forEachIndexed { i, e -> assertSubset(tail[i].wire(), e, "$name: logs[$i]") }
        }
        (expect["logCount"] as? Number)?.let {
            assertEquals(it.toInt(), DSXLogBuffer.shared.recent().size, "$name: retained log total")
        }
        (expect["logsTail"] as? List<Map<String, Any?>>)?.let { exp ->
            val tail = DSXLogBuffer.shared.recent().takeLast(exp.size)
            exp.forEachIndexed { i, e -> assertSubset(tail[i].wire(), e, "$name: logsTail[$i]") }
        }
        (expect["callErrors"] as? List<Any?>)?.let {
            assertEquals(it.map(JSE::string), callErrors, "$name: callErrors")
        }
        for ((path, exp) in (expect["jseStore"] as? Map<String, Any?>) ?: emptyMap()) {
            val store = jseStore ?: error("$name: no jse step ran")
            val actual = JSE.eval(path, store, null)
            assertSubset(if (actual == NSNull) null else actual, exp, "$name: jseStore $path")
        }
    }
}
